File: fas2txt.php
#!/usr/bin/php -q

<?php
/*
 * fas2txt.php
 *
 * PHP script that parse FASTA files into the plain text format
 *
 * Copyright (C) 2015   Conrad Shyu (shyu4751@yahoo.com)
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later
 * version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program. If not, see <http://www.gnu.org/licenses/>.
 *
 * revised on July 26, 2006
 * revised on May 1, 2015
 * revised on June 17, 2020
 *
*/

define("DELIMITER", ";");

/*
 * note: there is not standard format for the annotation on fasta file
 * it is necessary to manually identify the annotation section and extract
 * required information
*/
class fasta
{
    private $key; private $dna;

    public function __construct($k, $s) {
        $this->dna = trim(str_replace(array("%", "_", "'", "&", ";", "\"", "\n"), "", $s));
        $this->key = trim($k);
    }   // default constructor

    public function __destruct() {
    }   // default destructor

    public function origin() {
        return($this->dna);
    }   // return the sequence

    public function length() {
        return(strlen($this->dna));
    }   // length of the string

    public function accession() {
        $t = explode(DELIMITER, $this->key);
        $a = explode(" ", $t[0]);
        return(trim($a[0]));
    }   // get the accession number

    public function locus() {
        $t = explode(DELIMITER, $this->key);
        return(trim($t[count($t) - 2]));
    }   // get the locus number

    public function strain() {
        $t = explode(DELIMITER, $this->key);
        return(trim($t[count($t) - 1]));
    }   // get the strain name

    public function organism() {
        return($this->strain());
    }   // get the strain name

    public function ambiguity() {
        return(substr_count($this->dna, "N"));
    }   // number of degenerate bases

    public function degenerate() {
        return($this->ambiguity());
    }   // number of degenerate bases

    public function reverse() {
        return(strrev($this->dna));
    }   // find the reverse of a string

    public function antisense() {
        return(strtr($this->dna, "ATUGCYRSWKMBDHVN", "TAACGRYSWMKVHDBN"));
    }   // find the complement of a string

    public function complement() {
        return($this->antisense());
    }   // find the complmenet of a string
}   // smart container implementation

/*
 * this file will load all fasta entries into memory
*/
function load($f) {
    $h = fopen($f, "r"); $list = array(); $fas = ""; $key = "";

    while (!feof($h)) {
        $b = trim(fgets($h, 8192), " \n\t\r\0\x0B");

        if (strlen($b) < 2) {
            continue;
        }   // skip comments and empty line

        if (!($b[0] == '>')) {
            $fas .= strtoupper($b); continue;
        }   // accumulate the sequence

        if (strlen($fas) > 0) {
            $list[$key] = $fas; $fas = "";
        }   // sequence is available

        $key = trim($b, ">"); $list[$key] = 0;
     }   // run through the entire file

    fclose($h); $list[$key] = $fas; return($list);
}   // end of load()

/*
 * main procedure starts here
*/
if ($argc < 3) {
    echo $argv[0] . ": fasta_file output_file\n";
    echo "\nThis PHP script will extract data from the FASTA file and\n";
    echo "convert them into the MiCA database format.\n";
    echo "\nThe parameters are:\n";
    echo " fasta_file - the FASTA file to be processed.\n";
    echo "              note: there are no error checking on the format.\n";
    echo "output_file - the name of the output file.\n";
    echo "For more information about this script, please consult a magician.\n";
    exit();
}   // print out the help

echo "loading file ... ";
$list = load($argv[1]); $db = array();
echo "completed\n";

echo "processing sequence(s) ... ";
foreach (array_keys($list) as $k) {
    $db[] = new fasta($k, $list[$k]);
}   // load the content into memory
echo "completed\n";

$h = fopen($argv[2], "w");

echo "writing database ... ";
foreach ($db as $i) {
    fprintf($h, "%s|%s|%s|%s\n", $i->organism(), $i->accession(), $i->locus(), $i->origin());
}   // write the database file
echo "completed\ntotal number of records: " . count($db) . "\n";

fclose($h);
?>

File: fas2txt.cs
/*
 * fas2txt.cs
 * This program converts sequence file in the fasta format into mica database.
 * Fasta format is generally not consistent, therefore it is necessary to update
 * the code in order to accommendate difference in the annotation section.
 *
 * Copyright (C) 2015   Conrad Shyu (shyu4751@yahoo.com)
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later
 * version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program. If not, see <http://www.gnu.org/licenses/>.
*/

/*
 * note: there is no standard format for the annotation on fasta file
 * it is necessary to manually identify the annotation section and extract
 * required information
*/
class fsa2txt
{
	/*
	 * this implementation loads the entire file into memory
	*/
    static System.Int32 fasta(
        System.Collections.Generic.Dictionary<System.String, System.String> _l,
        System.String _f )
    {
        System.String k = "";
        System.Text.StringBuilder s = new System.Text.StringBuilder();

        try
        {
            using ( System.IO.StreamReader r = new System.IO.StreamReader( _f ) )
            {
                System.String b = "";

                while ( ( b = r.ReadLine() ) != null )
                {
                    if ( b.Length < 2  )
                    {
                        continue;
                    }   // skip comments and empty line

                    if ( !( b[ 0 ] == '>' ) )
                    {
                        s.Append( b ); continue;
                    }   // accumulate the sequence

                    if ( ( s.ToString() ).Length > 0 )
                    {
                        _l[ k ] = ( s.ToString() ).Trim(); s.Length = 0;
                    }   // sequence is available

                    k = b.Trim( '>' ); _l.Add( k, "" );
                }   // set the taxonomic identification and rank
            }	// file is automatically closed
        }
        catch ( System.Exception e )
        {
            System.Console.WriteLine( "Exception: {0}", e.Message );
        }   // an exception has occurred

        _l[ k ] = ( s.ToString() ).Trim(); return( _l.Count );
    }   // extract sequences from fasta file

    public static int Main( System.String[] args )
    {
        if ( args.Length < 2 )
        {
            System.Console.WriteLine( "require: fasta output" ); return( 0 );
        }   // check the required parameters

        System.Collections.Generic.Dictionary<System.String, System.String> list =
            new System.Collections.Generic.Dictionary<System.String, System.String>();

        System.Console.Write( "processing: {0} ... ", args[ 0 ] );
        fasta( list, args[ 0 ] ); System.Console.WriteLine( "completed" );

        try
        {
            using ( System.IO.StreamWriter w = new System.IO.StreamWriter( args[ 1 ], false ) )
            {
                foreach( System.Collections.Generic.KeyValuePair<System.String, System.String> a in list )
                {
                    System.String[] id = ( a.Key ).Split( '|' );
                    w.WriteLine( "{0}|{1}|{2}|{3}",
					    ( ( ( id[ 4 ] ).Split( ',' ) )[ 0 ] ).Trim(),
                        ( id[ 1 ] ).Trim(), ( id[ 3 ] ).Trim(), ( a.Value ).Trim() );
                }   // process each entry
            }   // export the sequence into a file
        }
        catch ( System.Exception e )
        {
            System.Console.WriteLine( "Exception: {0}", e.Message );
        }   // an exception has occurred

        return( 1 ); 
    }
}
